api/src/main/java/org/apache/iceberg/expressions/UnboundPredicate.java

```diff
@@ -125,13 +154,38 @@ public Expression bind(Types.StructType struct, boolean caseSensitive) {
         case LT_EQ:
         case EQ:
           return Expressions.alwaysFalse();
-//        case IN:
-//          break;
-//        case NOT_IN:
-//          break;
       }
     }
     return new BoundPredicate<>(op(), new BoundReference<>(field.fieldId(),
-        schema.accessorForField(field.fieldId())), lit);
+            schema.accessorForField(field.fieldId())), lit);
+  }
+
+  @SuppressWarnings("unchecked")
+  private Expression bindInOperation(Types.NestedField field, Schema schema) {
+    final Set<Literal<T>> lits = literalSet().stream().map(
+        val -> {
+          if (val instanceof LiteralSet.CharSeqWrapper) {
+            val = (T) ((LiteralSet.CharSeqWrapper) val).unWrap();
+          }
+          Literal<T> lit = Literals.from(val).to(field.type());
+          if (lit == null) {
+            throw new ValidationException(String.format(
+                "Invalid value for comparison inclusive type %s: %s (%s)",
+                field.type(), val, val.getClass().getName()));
+          }
+          return lit;
+        })
+        .filter(l -> l != Literals.aboveMax() && l != Literals.belowMin())
+        .collect(Collectors.toSet());
```

Comment:
I don't think this needs to be a Set. That requires a `Literal#equals` implementation and doesn't really add any value. Instead, you could use `Iterable`:

```java
Iterable<Literal<T>> typedLits = Iterables.filter(
    Iterables.transform(literals, lit -> lit.to(field.type())),
    lit -> lit != Literals.aboveMax() && lit != Literals.belowMin());

if (Iterables.contains(typedLits, null)) {
  throw ...
}
```

Reply:
@rdblue The main reason to make it to be a `Set` is for deduplication.
For example,
```
StructType struct = StructType.of(required(15, "d", Types.DecimalType.of(9, 2)));
UnboundPredicate<Double> unbound = Expressions.in("d", 12.40, 12.401, 12.402);
Expression expr = unbound.bind(struct);
// expr should be `BoundPredicate<BigDecimal>` with `EQ` op.
```
So here, the size of literalSet might shrink due to the type conversion.

Reply:
There are several places where these values are deduplicated. I think it would be better if `UnboundPredicate` was always created with a `Collection<Literal<T>>` (that is a set when created by `in` and `notIn`) and stored as an immutable list. That would simplify tracking literals because we can use the same storage for null/notNull (empty list), normal predicates (list of one value), and in/notIn (list of values).

Reply:
Thanks for the suggestion. Updated the code accordingly.